src/op_handlers/fat_pointer_read.rs

```diff
@@ -10,6 +10,13 @@ pub fn fat_pointer_read(vm: &mut VMState, opcode: &Opcode) -> Result<(), EraVmEr
     if !src0.is_pointer {
         return Err(OperandError::InvalidSrcNotPointer(opcode.variant).into());
     }
+
+    // make sure the pointer is a value that can be added without overflow for 32 bits.
+    if src0.value.low_u32() > (u32::MAX - 32) {
+        let _ = vm.set_gas_left(0);
```

Comment:
Handle error:
```rs
vm.set_gas_left(0)?;
```

Reply:
In their implementation, they burn all gas as well. See [here](https://github.com/matter-labs/vm2/blob/master/src/instruction_handlers/heap_access.rs#L68).

Reply:
We don't return the err, because we will return an err anyways, and we want to specify that it is because the address was invalid

Reply:
Burning all gas seems correct.

Their [`use_gas(&mut self, amount: u32)`](https://github.com/matter-labs/vm2/blob/9a38900d7af9b1d72b47ce3be980e77c1239a61d/src/state.rs#L82-L83) function will only return an `Err` if `self.current_frame.gas < amount`, and even then it will set the current gas to zero, that's why they don't need to handle the error.

Our gas setting function may also fail due to a `Vec<>` access error (see `VMState::current_context_mut`).

Reply:
You are right!